dbos/system_database.go

```diff
@@ -671,6 +728,184 @@ func (s *sysDB) cancelWorkflow(ctx context.Context, workflowID string) error {
+		loadInput:   true,
+		loadOutput:  true,
+		tx:          tx,
+	}
+	wfs, err := s.listWorkflows(ctx, listInput)
+	if err != nil {
+		s.logger.Error("ResumeWorkflow: failed to list workflows", "error", err)
+		return err
+	}
+	if len(wfs) == 0 {
+		return newNonExistentWorkflowError(workflowID)
+	}
+
+	wf := wfs[0]
+	if wf.Status == WorkflowStatusSuccess || wf.Status == WorkflowStatusError {
+		return nil // Workflow is complete, do nothing
+	}
+
+	// Set the workflow's status to ENQUEUED and clear its recovery attempts and deadline
+	updateStatusQuery := `UPDATE dbos.workflow_status
+						  SET status = $1, queue_name = $2, recovery_attempts = $3, 
+						      workflow_deadline_epoch_ms = NULL, deduplication_id = NULL, 
+						      started_at_epoch_ms = NULL, updated_at = $4
+						  WHERE workflow_uuid = $5`
+
+	_, err = tx.Exec(ctx, updateStatusQuery,
+		WorkflowStatusEnqueued,
+		_DBOS_INTERNAL_QUEUE_NAME,
+		0,
+		time.Now().UnixMilli(),
+		workflowID)
+	if err != nil {
+		return fmt.Errorf("failed to update workflow status to ENQUEUED: %w", err)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+type forkWorkflowDBInput struct {
+	originalWorkflowID string
+	forkedWorkflowID   string
+	startStep          int
+	applicationVersion string
+}
+
+func (s *sysDB) forkWorkflow(ctx context.Context, input forkWorkflowDBInput) error {
+	// Validate startStep
+	if input.startStep < 0 {
+		return fmt.Errorf("startStep must be >= 0, got %d", input.startStep)
+	}
+
+	tx, err := s.pool.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	// Get the original workflow status
+	listInput := listWorkflowsDBInput{
+		workflowIDs: []string{input.originalWorkflowID},
+		loadInput:   true,
+		loadOutput:  true,
+		tx:          tx,
+	}
+	wfs, err := s.listWorkflows(ctx, listInput)
+	if err != nil {
+		return fmt.Errorf("failed to list workflows: %w", err)
+	}
+	if len(wfs) == 0 {
+		return newNonExistentWorkflowError(input.originalWorkflowID)
+	}
+
+	originalWorkflow := wfs[0]
+
+	// Check if the workflow has completed successfully (required for forking)
+	if originalWorkflow.Status != WorkflowStatusSuccess {
+		return fmt.Errorf("cannot fork workflow %s: workflow must be in SUCCESS state, current state: %s", input.originalWorkflowID, originalWorkflow.Status)
+	}
+
+	// Validate that startStep doesn't exceed the workflow's actual steps
+	maxStepQuery := `SELECT COALESCE(MAX(function_id), 0) FROM dbos.operation_outputs WHERE workflow_uuid = $1`
+	var maxStepID int
+	err = tx.QueryRow(ctx, maxStepQuery, input.originalWorkflowID).Scan(&maxStepID)
+	if err != nil {
+		return fmt.Errorf("failed to query max step ID: %w", err)
+	}
+	if input.startStep > maxStepID && maxStepID > 0 {
+		return fmt.Errorf("startStep %d exceeds workflow's maximum step %d", input.startStep, maxStepID)
+	}
```

Comment:
We might not need this check. I think it's useful for the user to know that their requested step doesn't exist. The Python behavior defaults to copying "steps up to N"

Reply:
We don't need this check, it's complex and doesn't add much